### DISCUSSION TITLE
Deferrable Tasks do not Respect Task Pools

### DISCUSSION THREAD

**Discussion by ReadytoRocc** [apache/airflow#21243] [General]
### Apache Airflow version

2.2.3 (latest released)

### What happened

When using a deferrable operator, the pool slot is released after the task goes into a state of deferred.

### What you expected to happen

I would expect the task to retain the pool slot for the entire execution across the worker and triggerer resources, and only release the pool once it reaches a terminal state.

### How to reproduce

Create a DAG the below dag, and a Pool named `async` with a slot count of `1`. Let this dag run, and notice that tasks first 16 tasks enter `queued` -> `running` -> `deferred`.

```
from datetime import datetime
from airflow import DAG
from airflow.sensors.date_time import DateTimeSensorAsync

with DAG(
    "async_dag",
    start_date=datetime(2022, 1, 18),
    schedule_interval="* * * * *",
    catchup=True,
) as dag:

    async_sensor = DateTimeSensorAsync(
        task_id="async_task",
        target_time="""{{ macros.datetime.utcnow() + macros.timedelta(minutes=20) }}""",
        pool="async",
    )
```

### Operating System

Debian GNU/Linux 11 (bullseye)

### Deployment

Other Docker-based deployment

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
I think this is an intended feature, not a bug. 

It's the matter of assumption/definiton of whether the deferred task takes the pool slot or not. I think (but I would love to hear what others think) this is quite a reasonable assumption that deferred task does not take a slot.

Taking into account that the deferred task - by definition takes almost no resources and any external communication is asynchronous and pool feature is designed to limit resources an potentially decrease the loaad/throttle external, mostly synchronous services, I think not taking a pool slot by deferred task is quite reasonable.

Does the task take the slot back when it is back from defferred state @ReadytoRocc ?

**Comment by ReadytoRocc** [comment]
> Does the task take the slot back when it is back from defferred state?

@potiuk yes. Once the `trigger` is finished, the task will be marked as `SCHEDULED`. The task will then re-queue once a `pool` slot is available.

**Comment by potiuk** [comment]
So I'd say it's pretty expected behaviour. I wonder what others think.

**Comment by potiuk** [comment]
@andrewgodwin @uranusjr  - WDYT ? I think the behaviour is intended and maybe just need a bit clarification in the docs.

**Comment by andrewgodwin** [comment]
This was indeed originally by design, as I saw pools as a "worker/active task slot" system, designed to concurrency-limit things that were running - and of course, deferred tasks are not really running.

I think either option here is reasonable, and we could change it to have deferred tasks use up a pool slot if that's the consensus, but I went with this behaviour in the initial feature as it seemed most "sensible". It's probably worth documenting it explicitly, though?

**Comment by potiuk** [comment]
I agree with @andrewgodwin @ReadytoRocc - is that explanation/behaviour plausible for you ?

**Comment by potiuk** [comment]
In short:

![image](https://user-images.githubusercontent.com/595491/150692540-618e1ad5-04a2-45fc-bd79-b80766a7fdf8.png)

**Comment by ReadytoRocc** [comment]
@andrewgodwin & @potiuk thank you for the detailed responses and for confirming this behavior is expected. Are we able to convert this `issue` into a `discussion`? I believe there are some use cases for using pools with deferrable operators.

An example would be if you want to limit the use of an external service that Airflow is orchestrating. Another would be to limit certain sets of tasks from using all of the triggerer’s resources.

Please let me know your thoughts, thanks!

**Reply by potiuk** [reply]
Can you give some specific examples? I think async operations almost by definitions impose pretty much no limitations on external systems. You cannot make a "traditional" hanging http:// request as Trigger - you'd have to use websocket for that and I hardly imagine there is an external service with websockets that would impose some limits .

**Reply by ReadytoRocc** [reply]
Of course. An example would be to limit the number of data processing jobs that can concurrently run on a Spark Cluster. Using a pool, we can ensure that Airflow only starts/runs `N` jobs (tasks) on that system at a time.

**Reply by andrewgodwin** [reply]
Yes, I would agree that external resources are probably the primary use-case for this. I think another open question is - if we were to make Deferred tasks take up space in a pool, would that be just on all the time, or opt-in on a per-task or per-pool basis? I don't want to make the code more complex, but I also don't want to go changing this behaviour by default on people.

**Reply by potiuk** [reply]
I think being able to specify "pool behaviour" on deferral would be great. something like `free_pool_on_defer` parameter - default 'true' ? Sound like an easy thing to pass as a the `TaskDeffered` and "defer" method?

**Reply by potiuk** [reply]
It does add some complexity though - new column in the DB and handling on resume/cancel etc.

**Reply by kmehkeri** [reply]
I think that pools and what @ReadytoRocc would like to have work slightly differently and we might want to have both at the same time.

My understanding is that pools are a way to throttle processes _at Airflow backend_. So the way it works now makes sense to me as well.

But some time ago I was also looking for resources/locks in Airflow that would allow to control concurrency or limit access to _external resources_. I would really like to be able to say:

```python
BashOperator(
    task_id="my_task",
    bash_command="echo hello",
    pool_slots=2,
    pool="my_pool",
    resources={"r1": 1, "r2": 3}
)
```

Above task would require 1 unit of r1 resource and 3 units of r2 resource. Unlike pools, these resources would be kept acquired as long as the task runs, regardless of whether it's deferred or not at the moment. And we might be not limited to just one, a task could demand multiple resources, like in this example.

**Reply by potiuk** [reply]
Just to add. - this is maybe not very well known fact but we already have "pool_slots" in Airlfow that do just that - allows to take more pool slots by a task (but for single pool). 

I think it's a bad idea to control "external resources" by Airlfow. This is a very leaky approach. What happens if you start using the same "resources" outside of Airflow? Wha purpose it will serve? What if you start running two  or three airflow instances? 

If you want to control "external" resource usage - you should do it on the server side (and for example return  https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429).

**Reply by dylanbstorey** [reply]
> But some time ago I was also looking for resources/locks in Airflow that would allow to control concurrency or limit access to external resources.

While i appreciate the wish to be respectful of external processing systems, it seems that the implementation of what is essentially a resource throttling / queueing system should be handled by those systems and not be something Airflow is responsible for (So i guess i'm agreeing with @potiuk  here that this is something the external service provider should manage)

**Reply by kmehkeri** [reply]
Yes, resource control would only apply to tasks ran via Airflow and if anything (manual action, other Airflow instance, ...) reaches the external service it will bypass limits. I can think of two reasons why - being aware of drawbacks - I would still like to use the feature:
* Sometimes I just don't control that service to enforce throttling on its side
* Tasks that are running, but held by external system's scheduling/throttling, still occupy Airflow worker's resources; instead they could stay queued in Airflow waiting for their turn

If you hold up that this should not be responsibility of Airflow, that's okay, there are always some workarounds possible. But if it could be accepted, I can try to contribute this myself.

**Reply by potiuk** [reply]
> Tasks that are running, but held by external system's scheduling/throttling, still occupy Airflow worker's resources; instead they could stay queued in Airflow waiting for their turn

This is not really true for DeferrableTasks. When they are deferred they don't take the resources almost by definition (or at least very minimal resources as determined by async.io). This is the whole point of the async.io framework that for all practical purposes it should allow you to run pretty much infinite number of tasks waiting for something. The main purpose of async.io is to make it possible and the only reason for implementing throttling is (or at least this is how I see it) is that non-async connected clients take precious resources from the server.

I know it's an over-simplification, but well, If you do expect limited number of your clients (with some reasonable limits counted in say - 10s or 100s ), why would you want to make the task deferred in the first place?

**Reply by kmehkeri** [reply]
> This is not really true for DeferrableTasks

Yes, that's valid for some cases. For example, I am already looking towards deferrable Spark operator to submit the task to Yarn, which would decide by its own rules when to run the task - and I don't need to worry about throttling/pooling from Airflow side. But this works if - and only if - remote system has a scheduler and API to poll about the result.

One case that I also had is orchestrating jobs to run on a set of some old VMs, via SSH/WinRM. There are some limitations about how many jobs of given type can run, but that's in no way checked, so I control this at scheduler level. With Airflow, pools could be used here with some limitations, but not for all cases (sometimes I want to limit on more than 1 resource).

Another case was syncing ETL jobs so that writes and reads on tables don't clash, especially where database engine has a poor support for locking (hello, big data).

**Reply by kaxil** [reply]
>Yes, that's valid for some cases. For example, I am already looking towards deferrable Spark operator to submit the task to Yarn, which would decide by its own rules when to run the task - and I don't need to worry about throttling/pooling from Airflow side. But this works if - and only if - remote system has a scheduler and API to poll about the result.

Yes, the system needs to support polling and most of them in our investigation do. For example, we at Astronomer have built a [`LivyOperatorAsync`](https://github.com/astronomer/astronomer-providers/blob/main/astronomer/providers/apache/livy/example_dags/example_livy.py#L26-L35) that uses Livy to submit a spark job and poll for it.

**Reply by preethamrajk** [reply]
@potiuk 

In our case, we orchestrate ADF pipelines using Airflow and we have used pools to limit the usage of the systems that the ADF connects to - some on-prem sources etc.

Now, we are planning on updating our operators to Deferred Operators - mainly because these are long running ADF Pipelines and ADF Operators are perfect to reduce the load on airflow workers (which are not doing any heavy-lifting anyway). But, this means we lose the ability to control the load ADF puts on our on-prem systems. 

Any possibility to allow the usage of pools at a taskinstance level rather than at a worker level? Or any other ways to take care of this?

**Reply by preethamrajk** [reply]
@dylanbstorey 

![image](https://user-images.githubusercontent.com/16648593/189558779-8f7af2b1-b8b2-4b6e-b206-ead4e093579e.png)

https://airflow.apache.org/docs/apache-airflow/stable/concepts/pools.html#pools

Also, it does look like pools have been designed keeping "some systems" in mind.

**Reply by potiuk** [reply]
If you feel like adding support for pooled defferable operators - I think this is entirely possible for you to add. As explained - it's not implemented now - but if you have a need you (or if your company needs it, maybe it can sponsor such work or delegate someone to work on it)  could join to > 2100 people who  contributed to Airlfow when they missed a feature? 

I think this is the coolest way your company could pay back for the software they get for free :). If you need that feature, developing, testing it locally and contributig it back is a great idea.

**Reply by ReadytoRocc** [reply]
@potiuk - I have worked on some ideas based on this discussion, and would like to continue the work so I can contribute them. Do you recommend opening a new issue, or can we turn this discussion into a feature request?

**Reply by potiuk** [reply]
Ideally - turn it into PR and lead it to completion.

Feature requests have very little value here. They are often simply closed and generally there are no more than "wishlist". And if no-one volunteers to do it, it will be there, open and nothing will happen to it. The exsting if "feature" request does not make it any closer to implementation - but implementation in PR does. 

If you want to make a proposal that you would like to get comments, devlist thread is much better, especially if you want to propose a significant change in how airflow works. This is IMHO borderline - adding existing pool feature to deferred tasks/triggerrer makes sense and whilte not terribly necessary - seems like a natural progression of the "defferable" operators (also @andrewgodwin commented in the same tone). So I think it does not need neither AIP nor discussion in the devlist. 

So then, the next big think is make a PR to implement it. We do not need "issue" to create a PR. We are very light-weight and as-low-beauracracy as possible, so if you want to contribute a change .... just contribute it :).

The discussions in PR - even if it won't be yet complete but will show actual code, is almost always better than creating an issue describing what kind of code you are going to contribute. Worst thing that can happen - we will close the PR, best thing - it will get through reviews and get merged :).

**Reply by ReadytoRocc** [reply]
Thanks @potiuk for the guidance. I will create a ~~Feature~~ Pull Request 👍.

**Reply by dstandish** [reply]
While I understand the motivation, I don't really think we should add another state for this.
Look at how many states we have:
![image](https://user-images.githubusercontent.com/15932138/189984650-fa982584-b871-46f0-b096-a5f13bba2765.png)
It's kindof extreme already.  How is the airflow user supposed to remember which colors refer to what?  It takes up too much horizontal space on the UI as it is.
I think states need a refactor -- execution state vs resolution etc.
and, there's already a PR (https://github.com/apache/airflow/pull/26249) ready for merge that gives visibility into what's happening with your deferred task

BUT... just one person's opinion 🤷

**Reply by dstandish** [reply]
maybe we benefit from something like the "[pod phase](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase)" of kubernetes.... a high level summary of task state.... not intended to be a fully comprehensive state machine.... then we can go wild with detailed statuses to freely support the workings of the scheduler etc with worrying about kludging up the UI and UX...

**Comment by notatallshaw-gts** [comment]
FYI I am relying on the behavior that given an async operator the trigger task doesn't take up a pool slot but the worker processes does.

We have occasional performance issue with lots of processes starting at once (e.g. say we want to start 50 FileSensorAsyncs), but we don't have any performance issues with the triggerer running many trigger tasks. So we put these async operators under a pool and it limits too many starting or stopping all at once.

Given the different runtime context of trigger tasks vs. worker processes I feel that if pools applied to both I would want a way to distinguish between them and get different behavior (like we do now)

**Reply by potiuk** [reply]
Any proposals -  PR that can be disucssed in detail on code proposal, are most welcome.

**Comment by Usiel** [comment]
I added a PR at https://github.com/apache/airflow/pull/32709 that works for me in some basic testing. It introduces a new flag on `Pool` that allows the admin to decide whether to include deferred tasks in the occupied pool slots count.
